From 740efddf69b264d6439f9cf124d36acab812dd39 Mon Sep 17 00:00:00 2001
From: Hector Martin <marcan@marcan.st>
Date: Sat, 30 Jul 2022 16:16:45 +0900
Subject: [PATCH 134/171] dockchannel-hid: Handle M2 Air keyboard types

It looks like the keyboard type isn't just for the overall layout kind,
but also varies by device. Conjecture is that Apple always allocates
them sequentially in groups of 3 for the 3 available physical layout
groups, so just do % 3.

Signed-off-by: Hector Martin <marcan@marcan.st>
---
 drivers/hid/dockchannel-hid/dockchannel-hid.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/hid/dockchannel-hid/dockchannel-hid.c b/drivers/hid/dockchannel-hid/dockchannel-hid.c
index aeaeb60edf39..ea9d57ed9fa4 100644
--- a/drivers/hid/dockchannel-hid/dockchannel-hid.c
+++ b/drivers/hid/dockchannel-hid/dockchannel-hid.c
@@ -615,8 +615,12 @@ static void dchid_create_interface_work(struct work_struct *ws)
 	} else if (!strcmp(iface->name, "keyboard")) {
 		hid->type = HID_TYPE_SPI_KEYBOARD;
 
-		/* These country codes match what earlier Apple HID keyboards did */
-		switch (dchid->device_id.keyboard_type) {
+		/*
+		 * These country codes match what earlier Apple HID keyboards did.
+		 * Apple seems to allocate keyboard IDs in groups of 3 (for the 3
+		 * layout groups), hence the % 3.
+		 */
+		switch (dchid->device_id.keyboard_type % 3) {
 		case KEYBOARD_TYPE_ANSI:
 			hid->country = 33; // US-English
 			break;
-- 
2.34.1

